File: src/main/java/dev/todaka/robustredis/RedisCommands.java
package dev.todaka.robustredis;

import dev.todaka.robustredis.protocol.*;

import java.util.concurrent.CompletableFuture;

/**
 * See https://redis.io/commands
 */
public interface RedisCommands extends CommandDispatcher {
    // https://redis.io/commands#cluster

    default CompletableFuture<String> clusterNodes() {
        final var input = new CommandInput(CommandName.CLUSTER).addArg("NODES");
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }

    // https://redis.io/commands#connection

    default CompletableFuture<String> echo(String message) {
        final var input = new CommandInput(CommandName.ECHO).addArg(message);
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }

    default CompletableFuture<String> ping() {
        final var input = new CommandInput(CommandName.PING);
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }

    // https://redis.io/commands#generic

    default CompletableFuture<String> del(String key) {
        final var input = new CommandInput(CommandName.DEL).addKey(key);
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }

    default CompletableFuture<Long> exists(String key) {
        final var input = new CommandInput(CommandName.EXISTS).addKey(key);
        return dispatchCommand(new RedisCommand<>(input, new LongCommandOutput()));
    }

    // https://redis.io/commands#string

    default CompletableFuture<Long> decr(String key) {
        final var input = new CommandInput(CommandName.DECR).addKey(key);
        return dispatchCommand(new RedisCommand<>(input, new LongCommandOutput()));
    }

    default CompletableFuture<String> get(String key) {
        final var input = new CommandInput(CommandName.GET).addKey(key);
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }

    default CompletableFuture<Long> incr(String key) {
        final var input = new CommandInput(CommandName.INCR).addKey(key);
        return dispatchCommand(new RedisCommand<>(input, new LongCommandOutput()));
    }

    default CompletableFuture<String> set(String key, String value) {
        final var input = new CommandInput(CommandName.SET).addKey(key).addArg(value);
        return dispatchCommand(new RedisCommand<>(input, new StringCommandOutput()));
    }
}

File: src/main/kotlin/dev/todaka/robustredis/RedisCommands.kt
package dev.todaka.robustredis

import dev.todaka.robustredis.model.*
import java.util.concurrent.CompletableFuture

interface CommandDispatcher {
    fun <R> dispatchCommand(redisCommand: RedisCommand<R>): CompletableFuture<R>
}

interface RedisCommands : CommandDispatcher {
    // https://redis.io/commands#cluster
    fun clusterNodes(): CompletableFuture<String> {
        val input = CommandInput(CommandName.CLUSTER, args = listOf("NODES"))
        return dispatchCommand(RedisCommand(input, StringCommandOutput()))
    }

    // https://redis.io/commands#connection
    fun echo(message: String): CompletableFuture<String> {
        val input = CommandInput(CommandName.ECHO, args = listOf(message))
        return dispatchCommand(RedisCommand(input, StringCommandOutput()))
    }

    fun ping(): CompletableFuture<String> {
        val input = CommandInput(CommandName.PING)
        return dispatchCommand(RedisCommand(input, StringCommandOutput()))
    }

    // https://redis.io/commands#generic
    fun del(key: String): CompletableFuture<Long> {
        val input = CommandInput(CommandName.DEL, keys = listOf(key))
        return dispatchCommand(RedisCommand(input, LongCommandOutput()))
    }

    fun exists(key: String): CompletableFuture<Long> {
        val input = CommandInput(CommandName.EXISTS, keys = listOf(key))
        return dispatchCommand(RedisCommand(input, LongCommandOutput()))
    }

    // https://redis.io/commands#string
    fun decr(key: String): CompletableFuture<Long> {
        val input = CommandInput(CommandName.DECR, keys = listOf(key))
        return dispatchCommand(RedisCommand(input, LongCommandOutput()))
    }

    fun get(key: String): CompletableFuture<String?> {
        val input = CommandInput(CommandName.GET, keys = listOf(key))
        return dispatchCommand(RedisCommand(input, NullableStringCommandOutput()))
    }

    fun incr(key: String): CompletableFuture<Long> {
        val input = CommandInput(CommandName.INCR, keys = listOf(key))
        return dispatchCommand(RedisCommand(input, LongCommandOutput()))
    }

    fun set(key: String, value: String): CompletableFuture<String> {
        val input = CommandInput(CommandName.SET, keys = listOf(key), args = listOf(value))
        return dispatchCommand(RedisCommand(input, StringCommandOutput()))
    }
}
